iree/compiler/Dialect/Flow/Transforms/CreateFuncsToInvokeExecOps.cpp

```diff
@@ -0,0 +1,84 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iree/compiler/Dialect/Flow/IR/FlowOps.h"
+#include "iree/compiler/Dialect/IREE/IR/IREEOps.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/IR/BlockAndValueMapping.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/StandardTypes.h"
+#include "mlir/Pass/Pass.h"
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace Flow {
+
+// Walks through all the execuatable ops and creates Funcs to invoke them. The
+// input are provided using iree.unfoldable_constant.
+class CreateFuncsToInvokeExecOpsPass
+    : public PassWrapper<CreateFuncsToInvokeExecOpsPass,
+                         OperationPass<ModuleOp>> {
+ public:
+  CreateFuncsToInvokeExecOpsPass() = default;
+
+  void runOnOperation() override {
+    ModuleOp moduleOp = getOperation();
+    auto builder = OpBuilder::atBlockBegin(moduleOp.getBody());
+    Location loc = moduleOp.getLoc();
+    auto execOps = moduleOp.getOps<IREE::Flow::ExecutableOp>();
+    for (auto execOp : execOps) {
+      for (auto& op : execOp.getBlock()) {
+        if (auto dispatchEntryOp = dyn_cast<IREE::Flow::DispatchEntryOp>(op)) {
+          auto execFuncOp = execOp.getInnerModule().lookupSymbol<FuncOp>(
+              dispatchEntryOp.function_ref());
+          std::string funcName = std::string(execFuncOp.getName()) + "_entry";
+          auto funcType =
+              builder.getFunctionType({}, execFuncOp.getType().getResults());
+          auto funcOp =
+              builder.create<FuncOp>(moduleOp.getLoc(), funcName, funcType);
+          funcOp.setAttr("iree.module.export", UnitAttr::get(&getContext()));
+          Block* block = funcOp.addEntryBlock();
+          auto blockBuilder = OpBuilder(block, block->begin());
+          SmallVector<Value, 4> args;
+          for (auto inputType : execFuncOp.getType().getInputs()) {
+            // TODO(hanchung): Use non-zero or random values as inputs.
+            auto attr = blockBuilder.getZeroAttr(inputType);
+            auto cst = blockBuilder.create<ConstantOp>(moduleOp.getLoc(),
+                                                       inputType, attr);
+            args.push_back(
```

Comment:
instead of this (do not optimize) you can run createOutlineLargeConstantsPass immediately after this in your pipeline to move the constants to flow.variables; this will ensure that you get proper constant handling in the following HAL pipeline (vs this, which will be allocating and filling these buffers at runtime in your benchmark, which is not what you want)

example: https://github.com/google/iree/blob/e96c7f16a40ebf943f50c7069d528c9b72c64dde/iree/compiler/Dialect/Flow/Transforms/test/outline_large_constants.mlir#L3-L8

Reply:
Done, thanks for pointing this out!

One question, I still don't get the idea of moving them to flow.variables. I think at some point the HAL should pass this information to the kernel, but how can it do it without any allocating and filling? This is so magic! Don't we need a place or buffer to store flow.variables during execution?

Reply:
variables can be immutable (constants) or mutable (where constants can be used as initializers), and both types get handled by the constant pooling logic:
https://github.com/google/iree/blob/44fd39522633694b30892b68fa3568d3b41c8250/iree/compiler/Dialect/HAL/Transforms/test/identify_constant_pools.mlir#L3-L40

splats get turned into special hal.constant_pool.splat values:
https://github.com/google/iree/blob/44fd39522633694b30892b68fa3568d3b41c8250/iree/compiler/Dialect/HAL/Transforms/test/pack_constant_pool_storage.mlir#L10-L11

and then when the runtime storage buffers are materialized the splats get turned into alloc+fill (vs dense constants, which are uploaded differently):
https://github.com/google/iree/blob/44fd39522633694b30892b68fa3568d3b41c8250/iree/compiler/Dialect/HAL/Transforms/test/materialize_constant_pool_buffers.mlir#L19-L31

So at runtime you'll have a !hal.buffer with the constant value, whether it was created with a splat fill or dense upload. In this way besides how the buffer is created the behavior between dense and splat values is identical at runtime (they are both fully-sized buffers).

If you don't go down the variable/constant pooling path then today you'll be allocating this buffer every call and not reusing it. In the future there are valid use cases of having constants locally in command buffers (less peak memory usage for small transient constants, or transient constants that are determined themselves at runtime like channel masking tensors). The do_not_optimize, though, will prevent those cases from working correctly and is something we should never really be adding unless it's a handwritten test case.